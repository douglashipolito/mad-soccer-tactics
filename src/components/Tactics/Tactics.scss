.tactics-component {
  .players-list-container {
    color: #000;
    padding: 10px;
    text-align: center;
    box-shadow: 0px 2px 5px 1px rgba(0, 0, 0, 0.5);
  }

  .team-1-list {
    border: 3px double #0344af;
    margin-right: -15px;
    .team-name {
      color: #0344af;
    }
  }

  .team-2-list {
    border: 3px double #d83c3c;
    margin-left: -15px;

    .team-name {
      color: #d83c3c;
    }
  }
}
