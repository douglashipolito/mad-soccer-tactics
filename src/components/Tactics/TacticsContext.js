import React from "react";

const TacticsContext = React.createContext({
  teams: [],
  changePlayerPosition() {},
  handleDragStart() {},
  handleDragEnter() {},
  handleDragLeave() {},
  handleDragOver() {},
  handleDragEnd() {},
  handleDrop() {},
  handleTacticChange() {},
  tactics: []
});

export const Provider = TacticsContext.Provider;
export const Consumer = TacticsContext.Consumer;
