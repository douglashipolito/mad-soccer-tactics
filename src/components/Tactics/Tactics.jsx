import React from "react";
import { Provider } from "./TacticsContext";
import PlayersList from "../PlayersList";
import Field from "../Field";
import teamsData from "../../../data/teams.json";
import tacticsData from "../../../data/tactics.json";
import "./Tactics.scss";
import TacticSelector from "../TacticSelector";

class Tactics extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      teams: [],
      changePlayerPosition: this.changePlayerPosition,
      handleDragStart: this.handleDragStart,
      handleDragEnter: this.handleDragEnter,
      handleDragLeave: this.handleDragLeave,
      handleDragOver: this.handleDragOver,
      handleDrop: this.handleDrop,
      handleDragEnd: this.handleDragEnd,
      tactics: tacticsData,
      handleTacticChange: this.handleTacticChange
    };
  }

  componentDidMount() {
    this.setTeamsTactics();
  }

  handleTacticChange = (event, teamId) => {
    const tactic = this.state.tactics.filter(
      tactic => tactic.id === event.target.value
    )[0];

    const teams = this.state.teams.map(team => {
      if (team.id === teamId) {
        const previousTactic = team.tactic;
        team.tactic = Object.assign({}, tactic);
        team.tacticId = tactic.id;

        // Positioning back all players already positioned
        previousTactic.positions.forEach(previousPosition => {
          if (!previousPosition.playerId) {
            return;
          }

          team.tactic.positions.some(position => {
            if (position.index === previousPosition.index) {
              position.playerId = previousPosition.playerId;
              return true;
            }
          });
        });
      }

      return team;
    });

    this.setState({ teams });
  };

  setTeamPosition(team) {
    const positions = [];
    let playersIds = team.players.map(player => {
      player.selected = true;
      return player.id;
    });

    team.tactic.positions.forEach(position => {
      const newPosition = Object.assign({}, position);
      newPosition.playerId = playersIds.pop();
      positions.push(newPosition);
    });

    team.tactic.positions = positions;
  }

  setTeamsTactics() {
    const teams = teamsData.map(team => {
      const tactic = tacticsData.filter(
        tactic => tactic.id === team.tacticId
      )[0];

      team.tactic = Object.assign({}, tactic);
      return team;
    });

    this.setTeamPosition(teams[1]);
    this.setState({ teams });
  }

  isItemDraggable(teamId) {
    const team = this.state.teams.filter(team => team.id === teamId)[0];
    return !team ? false : team.configurable;
  }

  handleDragStart = (event, teamId, playerId) => {
    if (!this.isItemDraggable(teamId)) {
      event.preventDefault();
      return;
    }

    event.target.style.opacity = "0.5";
    event.dataTransfer.setData(
      "text/plain",
      JSON.stringify({ playerId, teamId })
    );
  };

  handleDragEnter = (event, teamId) => {
    if (!this.isItemDraggable(teamId)) {
      event.preventDefault();
      return;
    }

    event.target.classList.add("drag-over");
  };

  handleDragLeave = (event, teamId) => {
    if (!this.isItemDraggable(teamId)) {
      event.preventDefault();
      return;
    }

    event.target.classList.remove("drag-over");
  };

  handleDragOver = (event, teamId) => {
    if (!this.isItemDraggable(teamId)) {
      return;
    }

    event.preventDefault();
    event.dataTransfer.dropEffect = "move";
  };

  handleDrop = (event, teamId, positionId) => {
    const data = event.dataTransfer.getData("text/plain");

    if (!this.isItemDraggable(teamId) || !data) {
      return;
    }

    event.target.classList.remove("drag-over");
    event.target.classList.add("dropped");
    event.preventDefault();
    this.setPlayerPosition(JSON.parse(data), positionId);
  };

  handleDragEnd = event => {
    event.target.style.opacity = "1";
  };

  setPlayerPosition = (playerData, positionId) => {
    const team = this.state.teams.filter(
      team => team.id === playerData.teamId
    )[0];

    let previousPosition = null;

    // Reset any previous position
    team.tactic.positions.some(tacticPosition => {
      if (tacticPosition.playerId === playerData.playerId) {
        previousPosition = tacticPosition;
        tacticPosition.playerId = null;
        return true;
      }
    });

    team.tactic.positions.some(tacticPosition => {
      if (tacticPosition.id === positionId) {
        if (tacticPosition.playerId && previousPosition) {
          previousPosition.playerId = tacticPosition.playerId;
        }

        tacticPosition.playerId = playerData.playerId;
        return true;
      }
    });

    team.players.some(player => {
      if (player.id === playerData.playerId) {
        player.selected = true;
        return true;
      }
    });

    this.setState({
      team
    });
  };

  render() {
    return (
      <div className="container tactics-component">
        <Provider value={this.state}>
          <div className="col-2 players-list-container team-1-list">
            <TacticSelector
              tactics={this.state.tactics}
              team={this.state.teams[0]}
            />
            <PlayersList team={this.state.teams[0]} />
          </div>
          <div className="col-8 field-container">
            <Field />
          </div>
          <div className="col-2 players-list-container team-2-list">
            <TacticSelector
              tactics={this.state.tactics}
              team={this.state.teams[1]}
            />
            <PlayersList team={this.state.teams[1]} />
          </div>
        </Provider>
      </div>
    );
  }
}

export default Tactics;
