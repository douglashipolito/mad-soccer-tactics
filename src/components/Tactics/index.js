import Tactics from "./Tactics.jsx";

export default Tactics;
