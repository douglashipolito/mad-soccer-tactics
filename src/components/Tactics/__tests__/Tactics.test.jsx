import React from "react";
import { create } from "react-test-renderer";
import Tactics from "../";

test("Tactics Renders", () => {
  const string = create(<Tactics />);
  expect(string).toMatchSnapshot();
});
