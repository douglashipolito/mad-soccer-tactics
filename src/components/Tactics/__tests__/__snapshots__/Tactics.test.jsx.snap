// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tactics Renders 1`] = `
<div
  className="container tactics-component"
>
  <div
    className="col-2 players-list-container team-1-list"
  >
    <div
      className="tactice-selector-component"
    >
      <form>
        <label
          htmlFor="tactic-selector-t1"
        >
          <span>
            Tactic
          </span>
          <select
            onChange={[Function]}
          >
            <option
              value="tactic1"
            >
              Tactic 1
            </option>
            <option
              value="tactic2"
            >
              Tactic 2
            </option>
            <option
              value="tactic3"
            >
              Tactic 3
            </option>
            <option
              value="tactic4"
            >
              Tactic 4
            </option>
          </select>
        </label>
      </form>
    </div>
    <section
      className="team-container player-list-container"
    >
      <h2
        className="team-name"
      >
        Blue Team
      </h2>
      <div
        className="player-list"
      >
        <ul
          className="team-list"
        >
          <li
            className="player"
          >
            <h3
              className="player-name"
            >
              <span>
                MarcusB
                 - 
                <small>
                  Tank
                </small>
              </span>
            </h3>
            <figure
              className="player-image-container"
              draggable={true}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <img
                alt="MarcusB"
                className="player-avatar"
                src="/images/chars/tank1.png"
              />
            </figure>
          </li>
          <li
            className="player"
          >
            <h3
              className="player-name"
            >
              <span>
                RafaF
                 - 
                <small>
                  Off-Tank
                </small>
              </span>
            </h3>
            <figure
              className="player-image-container"
              draggable={true}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <img
                alt="RafaF"
                className="player-avatar"
                src="/images/chars/tank2.png"
              />
            </figure>
          </li>
          <li
            className="player"
          >
            <h3
              className="player-name"
            >
              <span>
                EduM
                 - 
                <small>
                  DPS
                </small>
              </span>
            </h3>
            <figure
              className="player-image-container"
              draggable={true}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <img
                alt="EduM"
                className="player-avatar"
                src="/images/chars/dps1.png"
              />
            </figure>
          </li>
          <li
            className="player"
          >
            <h3
              className="player-name"
            >
              <span>
                KedidJ
                 - 
                <small>
                  Healer
                </small>
              </span>
            </h3>
            <figure
              className="player-image-container"
              draggable={true}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <img
                alt="KedidJ"
                className="player-avatar"
                src="/images/chars/healer.png"
              />
            </figure>
          </li>
        </ul>
      </div>
    </section>
  </div>
  <div
    className="col-8 field-container"
  >
    <div
      className="field-area"
    >
      <div
        className="football-pitch"
      >
        <ul
          className="positions team-1"
        >
          <li
            className="pos1 pos"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "left": "7%",
                "top": "20%",
              }
            }
          />
          <li
            className="pos2 pos"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "left": "75%",
                "top": "20%",
              }
            }
          />
          <li
            className="pos3 pos"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "left": "18%",
                "top": "60%",
              }
            }
          />
          <li
            className="pos4 pos"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "left": "68%",
                "top": "60%",
              }
            }
          />
        </ul>
        <ul
          className="positions team-2"
        >
          <li
            className="pos1 pos"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "left": "7%",
                "top": "20%",
              }
            }
          >
            <div
              className="player"
              draggable={true}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <div
                className="player-component"
              >
                <span
                  className="player-name"
                >
                  SelvaGi
                </span>
                <figure
                  className="player-image-container"
                >
                  <img
                    alt="SelvaGi"
                    className="img-responsive player-image"
                    src="/images/chars/healer2.png"
                  />
                </figure>
              </div>
            </div>
          </li>
          <li
            className="pos2 pos"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "left": "75%",
                "top": "20%",
              }
            }
          >
            <div
              className="player"
              draggable={true}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <div
                className="player-component"
              >
                <span
                  className="player-name"
                >
                  MaxiM
                </span>
                <figure
                  className="player-image-container"
                >
                  <img
                    alt="MaxiM"
                    className="img-responsive player-image"
                    src="/images/chars/tank4.png"
                  />
                </figure>
              </div>
            </div>
          </li>
          <li
            className="pos3 pos"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "left": "18%",
                "top": "60%",
              }
            }
          >
            <div
              className="player"
              draggable={true}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <div
                className="player-component"
              >
                <span
                  className="player-name"
                >
                  FoaG
                </span>
                <figure
                  className="player-image-container"
                >
                  <img
                    alt="FoaG"
                    className="img-responsive player-image"
                    src="/images/chars/tank3.png"
                  />
                </figure>
              </div>
            </div>
          </li>
          <li
            className="pos4 pos"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "left": "68%",
                "top": "60%",
              }
            }
          >
            <div
              className="player"
              draggable={true}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <div
                className="player-component"
              >
                <span
                  className="player-name"
                >
                  Goriathu
                </span>
                <figure
                  className="player-image-container"
                >
                  <img
                    alt="Goriathu"
                    className="img-responsive player-image"
                    src="/images/chars/tank1.png"
                  />
                </figure>
              </div>
            </div>
          </li>
        </ul>
        <div
          className="outline marking"
        />
        <div
          className="box left marking"
        />
        <div
          className="box-d left marking"
        />
        <div
          className="box left small marking"
        />
        <div
          className="box right marking"
        />
        <div
          className="box-d right marking"
        />
        <div
          className="box right small marking"
        />
        <div
          className="spot left marking"
        />
        <div
          className="spot right marking"
        />
        <div
          className="spot center marking"
        />
        <div
          className="center-line marking"
        />
        <div
          className="center-circle marking"
        />
        <div
          className="corner top left marking"
        />
        <div
          className="corner top right marking"
        />
        <div
          className="corner bottom left marking"
        />
        <div
          className="corner bottom right marking"
        />
        <div
          className="grass"
        />
      </div>
    </div>
  </div>
  <div
    className="col-2 players-list-container team-2-list"
  >
    <div
      className="tactice-selector-component"
    >
      <form>
        <label
          htmlFor="tactic-selector-t2"
        >
          <span>
            Tactic
          </span>
          <select
            onChange={[Function]}
          >
            <option
              value="tactic1"
            >
              Tactic 1
            </option>
            <option
              value="tactic2"
            >
              Tactic 2
            </option>
            <option
              value="tactic3"
            >
              Tactic 3
            </option>
            <option
              value="tactic4"
            >
              Tactic 4
            </option>
          </select>
        </label>
      </form>
    </div>
    <section
      className="team-container player-list-container"
    >
      <h2
        className="team-name"
      >
        Red Team
      </h2>
      <div
        className="player-list"
      >
        <ul
          className="team-list"
        >
          <li
            className="player selected"
          >
            <h3
              className="player-name"
            >
              <span>
                Goriathu
                 - 
                <small>
                  Tank
                </small>
              </span>
            </h3>
            <figure
              className="player-image-container"
              draggable={false}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <img
                alt="Goriathu"
                className="player-avatar"
                src="/images/chars/tank1.png"
              />
            </figure>
          </li>
          <li
            className="player selected"
          >
            <h3
              className="player-name"
            >
              <span>
                FoaG
                 - 
                <small>
                  Off-Tank
                </small>
              </span>
            </h3>
            <figure
              className="player-image-container"
              draggable={false}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <img
                alt="FoaG"
                className="player-avatar"
                src="/images/chars/tank3.png"
              />
            </figure>
          </li>
          <li
            className="player selected"
          >
            <h3
              className="player-name"
            >
              <span>
                MaxiM
                 - 
                <small>
                  DPS
                </small>
              </span>
            </h3>
            <figure
              className="player-image-container"
              draggable={false}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <img
                alt="MaxiM"
                className="player-avatar"
                src="/images/chars/tank4.png"
              />
            </figure>
          </li>
          <li
            className="player selected"
          >
            <h3
              className="player-name"
            >
              <span>
                SelvaGi
                 - 
                <small>
                  Healer
                </small>
              </span>
            </h3>
            <figure
              className="player-image-container"
              draggable={false}
              onDragEnd={[Function]}
              onDragLeave={[Function]}
              onDragStart={[Function]}
            >
              <img
                alt="SelvaGi"
                className="player-avatar"
                src="/images/chars/healer2.png"
              />
            </figure>
          </li>
        </ul>
      </div>
    </section>
  </div>
</div>
`;
